src/persist-client/src/impl/machine.rs

```diff
@@ -300,25 +289,91 @@ where
         }
     }
 
+    async fn maybe_init_state(
```

Comment:
I'm not entirely happy with this, so definitely looking for feedback if you have any. it seems almost but not quite redundant with `apply_unbatched_cmd` (apply_unbatched_cmd has a key difference in that after this commit it no longer has to worry about an un-initialized shard) plus the loop really shouldn't be necessary

Reply:
I'm afraid I don't yet follow where we use the fact that state is initialized to the latest state from consensus when creating the `Machine`. Most (all?) of the changes in the rest of the methods seem to come from moving the seqno to state?

Reply:
when seqno was on Machine, it was an Option<SeqNo> but now that it's on state, it's not an Option. all this effort for that...

I was being lazy squashing those two things (option -> not-option, machine -> state) into one commit but they could be separated if you think it'd be easier to review?

Reply:
I missed that one `unwrap_or_default()` call on `seqno` that we had in `apply_unbatched_command()` before... 🙈 

It's fine to leave the commits squashed as they are.

Reply:
Also, I agree that this and `apply_unbatched_cmd` seem very "duplicative", but I didn't come up with a suggestion for fixing it yet. It's probably fine to merge and revisit later when more things have settled.

Reply:
:thumbsup: